src/internal/minikubetestenv/client_factory.go

```diff
@@ -0,0 +1,102 @@
+package minikubetestenv
+
+import (
+	"context"
+	"sync"
+	"testing"
+
+	"github.com/pachyderm/pachyderm/v2/src/client"
+	"github.com/pachyderm/pachyderm/v2/src/internal/require"
+	"github.com/pachyderm/pachyderm/v2/src/internal/testutil"
+	"golang.org/x/sync/semaphore"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	poolSize        = 3
+	namespacePrefix = "test-cluster-"
+)
+
+var (
+	clusterFactory *ClusterFactory
+	setup          sync.Once
+)
+
+type ClusterFactory struct {
+	managedClusters   map[string]*client.APIClient
+	availableClusters map[string]bool
+	mu                sync.Mutex         // guards modifications to the ClusterFactory maps
+	sem               semaphore.Weighted // enforces max concurrency
+}
+
+func (cf *ClusterFactory) acquireFreeCluster() (string, *client.APIClient) {
+	cf.mu.Lock()
+	defer cf.mu.Unlock()
+	if len(cf.availableClusters) > 0 {
+		var assigned string
+		for ns := range cf.availableClusters {
+			assigned = ns
+			delete(cf.availableClusters, ns)
+			break
+		}
+		return assigned, cf.managedClusters[assigned]
+	}
+	return "", nil
+}
+
+func (cf *ClusterFactory) acquireNewCluster(t testing.TB) (string, *client.APIClient) {
+	assigned := testutil.UniqueString(namespacePrefix)
+	kube := testutil.GetKubeClient(t)
+	_, err := kube.CoreV1().Namespaces().Create(context.Background(),
+		&v1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: assigned,
+			},
+		},
+		metav1.CreateOptions{})
+	require.NoError(t, err)
+
+	cf.mu.Lock()
+	cf.managedClusters[assigned] = nil // hold my place in line
+	managedCount := len(cf.managedClusters)
+	cf.mu.Unlock()
+
+	c := InstallRelease(t,
+		context.Background(),
+		assigned,
+		kube,
+		&DeployOpts{
+			PortOffset: uint16(managedCount * 11),
+		})
+	cf.mu.Lock()
+	cf.managedClusters[assigned] = c
+	cf.mu.Unlock()
+	return assigned, c
+}
+
+// AcquireCluster returns a pachyderm APIClient from one of a pool of managed pachyderm
+// clusters deployed in separate namespace, along with the associated namespace
+func AcquireCluster(t testing.TB, ctx context.Context) (*client.APIClient, string) {
```

Comment:
Not sure how much value the context adds here.  We assume context.Backrgound() elsewhere.  Here, we use it to set the deadline for sem.Acquire.  I think this will be run at the beginning of every test, and in practice no one will want fine grained control over how long to wait for a cluster.  If it takes longer than 5 minutes, maybe something is wrong.  It's not something developers will be thinking about when they are writing tests.  I think we should make that decision (about how long to wait) here.

Reply:
sounds good to me - i agree 👍 I didn't put much thought behind use-cases here and thought it'd be nice to have.